### PR TITLE
Optimized MSMs for small scalars

### PR DESCRIPTION
**Pull request by moodlezoup**
Porting over some MSM optimizations from the jolt codebase
Reduces commitment time by 90% 🚀

### CODE REVIEW

File: src/provider/pasta.rs [hunk 1]
```diff
@@ -81,6 +81,26 @@ macro_rules! impl_traits {
...
+      }
+
+      fn msm_u64_wnaf(
+        _bases: &[Self::PreprocessedGroupElement],
+        _scalars: &[u64],
+        _max_num_bits: usize,
+      ) -> Self {
+        unimplemented!()
+      }
+
```
Comment (moodlezoup) [inline src/provider/pasta.rs:103]: we don't use pasta for Jolt and it feels gross to copy and paste everything here

Comment (sragss) [reply src/provider/pasta.rs:103]: 👍

File: src/provider/mod.rs [hunk 2]
```diff
@@ -38,7 +38,162 @@ macro_rules! impl_traits {
         scalars: &[Self::Scalar],
         bases: &[Self::PreprocessedGroupElement],
       ) -> Self {
-        best_multiexp(scalars, bases)
+        let max_num_bits = scalars
```
Comment (sragss) [inline src/provider/mod.rs:41]: Would be cheaper to construct G::Scalar::from(1 << {2, 10, 26}) and do direct `.lt() / .gt()` due to the cost of deconstructing a field element to bits (incurs a field multiplication to get out of montgomery form). 

This step may be cheap compared to the MSM, feel free to ignore if so.

**Review by sragss — APPROVED** [review]

File: src/provider/mod.rs (hunk 2 above)
Comment (moodlezoup) [reply src/provider/mod.rs:41]: it's very cheap, and we want to pass in `max_num_bits` to `msm_small` and `msm_u64_wnaf` anyway

**Pull request merged**